src/Draco.SourceGeneration/EquatableArray.cs

```diff
@@ -0,0 +1,106 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json.Serialization;
+using System.Text.Json;
+
+namespace Draco.SourceGeneration;
+
+internal static class EquatableArray
+{
+    public static EquatableArray<T> AsEquatableArray<T>(this ImmutableArray<T> array)
+        where T : IEquatable<T> => new(array);
+
+    public static EquatableArray<T> ToEquatableArray<T>(this IEnumerable<T> soruce)
+        where T : IEquatable<T> => new(soruce.ToImmutableArray());
+}
+
+[JsonConverter(typeof(EquatableArrayConverter))]
+internal readonly struct EquatableArray<T> : IEquatable<EquatableArray<T>>, IEnumerable<T>
```

Comment:
I'll add `IList<T>` (it might make LINQ marginally faster), but we can add other implementations when we actually need them. The implementations would just delegate to ImmutableArray anyway, so it's fairly simple boilerplate.